libraries/AC_PID/AC_P_1D.cpp

```diff
@@ -0,0 +1,171 @@
+//   the target is moved closer to the measurement and limit_min or limit_max will be set true
+float AC_P_1D::update_all(float &target, float measurement, bool &limit_min, bool &limit_max)
+{
+    limit_min = false;
+    limit_max = false;
+
+    // calculate distance _error
+    float error = target - measurement;
+
+    if (asymetric_limit(error, _lim_err_neg, _lim_err_pos, limit_min, limit_max )) {
+        target = measurement + error;
+    }
+
+    // ToDo: Replace sqrt_controller with optimal acceleration and jerk limited curve
+    // MIN(_Dxy_max, _D2xy_max / _kxy_P) limits the max accel to the point where max jerk is exceeded
+    return sqrt_controller(error, _kp, _lim_D_Out, _dt);
+}
+
+float AC_P_1D::get_p() const
+{
+    return _error * _kp;
+}
+
+void AC_P_1D::load_gains()
+{
+    _kp.load();
+}
+
+void AC_P_1D::save_gains()
+{
+    _kp.save();
+}
+
+/// Overload the function call operator to permit easy initialisation
+void AC_P_1D::operator()(float initial_p, float dt)
+{
+    _kp = initial_p;
+    _dt = dt;
+}
+
+/// set limits based on separate min and max
+bool AC_P_1D::asymetric_limit(float &input, float min, float max, bool &limit_min, bool &limit_max )
+{
+    if (input < min) {
+        input = min;
+        limit_min = true;
+    }
+    if (input > max) {
+        input = max;
+        limit_max = true;
+    }
+
+    return limit_min || limit_max;
+}
+
+// Proportional controller with piecewise sqrt sections to constrain second derivative
+float AC_P_1D::sqrt_controller(float error, float p, float D_max, float dt)
+{
+    float output;
+    if (!is_positive(D_max)) {
+        // second order limit is zero or negative.
+        output = error * p;
+    } else if (is_zero(p)) {
+        // P term is zero but we have a second order limit.
+        if (is_positive(error)) {
+            output = safe_sqrt(2.0f * D_max * (error));
+        } else if (is_negative(error)) {
+            output = -safe_sqrt(2.0f * D_max * (-error));
+        } else {
+            output = 0.0f;
+        }
+    } else {
+        // both the P and second order limit have been defined.
+        float linear_dist = D_max / sq(p);
+        if (error > linear_dist) {
+            output = safe_sqrt(2.0f * D_max * (error - (linear_dist / 2.0f)));
+        } else if (error < -linear_dist) {
+            output = -safe_sqrt(2.0f * D_max * (-error - (linear_dist / 2.0f)));
+        } else {
+            output = error * p;
+        }
+    }
+    if (!is_zero(dt)) {
+        // this ensures we do not get small oscillations by over shooting the error correction in the last time step.
+        return constrain_float(output, -fabsf(error) / dt, fabsf(error) / dt);
+    }
+
+    return output;
+}
+
+// Proportional controller with piecewise sqrt sections to constrain second derivative
+float AC_P_1D::inv_sqrt_controller(float output, float p, float D_max)
+{
+    output = fabsf(output);
+
+    float error;
+    if (!is_positive(D_max) && is_zero(p)) {
```

Comment:
I wonder if this would be clearer using early-returns...

Reply:
good idea, changed to use early returns